// @coreui/vue-pro/dist/cjs/components/tabs/CTabs.js
'use strict';

var vue = require('vue');
require('@popperjs/core');
var useUniqueId = require('../../composables/useUniqueId.js');

const CTabs = vue.defineComponent({
    name: 'CTabs',
    props: {
        /**
         * The active item key.
         */
        activeItemKey: {
            type: [Number, String],
            required: true,
        }
    },
    emits: [
        /**
         * The callback is fired when the active tab changes.
         */
        'change',
    ],
    setup(props, { slots, emit }) {
        const { getUID } = useUniqueId.useUniqueId();
        const uID = vue.ref(getUID());
        const activeItemKey = vue.ref(props.activeItemKey);
        const setActiveItemKey = (key) => {
            activeItemKey.value = key;
        };
        vue.watch(() => props.activeItemKey, (value) => {
            activeItemKey.value = value;
            emit('change', value);
        });
        vue.provide('activeItemKey', activeItemKey);
        vue.provide('id', uID);
        vue.provide('setActiveItemKey', setActiveItemKey);
        return () => vue.h('div', { class: 'tabs' }, slots.default && slots.default());
    },
});

exports.CTabs = CTabs;
