// @coreui/vue-pro/dist/cjs/components/tabs/CTab.js
'use strict';

var vue = require('vue');

const CTab = vue.defineComponent({
    name: 'CTab',
    props: {
        /**
         * Item key.
         */
        itemKey: {
            type: [Number, String],
            required: true,
        },
    },
    setup(props, { slots }) {
        const activeItemKey = vue.inject('activeItemKey');
        const id = vue.inject('id');
        const setActiveItemKey = vue.inject('setActiveItemKey');
        const isActive = () => props.itemKey === activeItemKey.value;
        return () => vue.h('button', {
            class: [
                'nav-link',
                {
                    active: isActive(),
                },
            ],
            id: `${props.itemKey}-tab-${id.value}`,
            role: 'tab',
            tabindex: isActive() ? 0 : -1,
            type: 'button',
            'aria-controls': `${props.itemKey}-tab-panel-${id.value}`,
            'aria-selected': isActive(),
            onClick: () => setActiveItemKey(props.itemKey),
            onFocus: () => setActiveItemKey(props.itemKey),
        }, slots.default && slots.default());
    },
});

exports.CTab = CTab;
