// @coreui/vue-pro/dist/cjs/components/accordion/index.js
'use strict';

var CAccordion = require('./CAccordion.js');
var CAccordionBody = require('./CAccordionBody.js');
var CAccordionButton = require('./CAccordionButton.js');
var CAccordionHeader = require('./CAccordionHeader.js');
var CAccordionItem = require('./CAccordionItem.js');

const CAccordionPlugin = {
    install: (app) => {
        app.component(CAccordion.CAccordion.name, CAccordion.CAccordion);
        app.component(CAccordionBody.CAccordionBody.name, CAccordionBody.CAccordionBody);
        app.component(CAccordionButton.CAccordionButton.name, CAccordionButton.CAccordionButton);
        app.component(CAccordionHeader.CAccordionHeader.name, CAccordionHeader.CAccordionHeader);
        app.component(CAccordionItem.CAccordionItem.name, CAccordionItem.CAccordionItem);
    },
};

exports.CAccordion = CAccordion.CAccordion;
exports.CAccordionBody = CAccordionBody.CAccordionBody;
exports.CAccordionButton = CAccordionButton.CAccordionButton;
exports.CAccordionHeader = CAccordionHeader.CAccordionHeader;
exports.CAccordionItem = CAccordionItem.CAccordionItem;
exports.CAccordionPlugin = CAccordionPlugin;
