// @coreui/vue-pro/dist/cjs/components/toast/CToastClose.js
'use strict';

var vue = require('vue');
var CCloseButton = require('../close-button/CCloseButton.js');

const CToastClose = vue.defineComponent({
    name: 'CToastClose',
    props: {
        /**
         * Component used for the root node. Either a string to use a HTML element or a component.
         */
        as: String,
        ...CCloseButton.CCloseButton.props,
    },
    emits: [
        /**
         * Event called before the dissmiss animation has started.
         */
        'close',
    ],
    setup(props, { slots, emit }) {
        // eslint-disable-next-line no-unused-vars
        const updateVisible = vue.inject('updateVisible');
        const handleClose = () => {
            emit('close');
            updateVisible(false);
        };
        return () => props.as
            ? vue.h(props.as, {
                onClick: () => {
                    handleClose();
                },
            }, () => slots.default && slots.default())
            : vue.h(CCloseButton.CCloseButton, {
                ...props,
                onClick: () => {
                    handleClose();
                },
            });
    },
});

exports.CToastClose = CToastClose;
