// @coreui/vue-pro/dist/cjs/components/form/index.js
'use strict';

var CForm = require('./CForm.js');
var CFormCheck = require('./CFormCheck.js');
var CFormFeedback = require('./CFormFeedback.js');
var CFormFloating = require('./CFormFloating.js');
var CFormInput = require('./CFormInput.js');
var CFormLabel = require('./CFormLabel.js');
var CFormRange = require('./CFormRange.js');
var CFormSelect = require('./CFormSelect.js');
var CFormSwitch = require('./CFormSwitch.js');
var CFormText = require('./CFormText.js');
var CFormTextarea = require('./CFormTextarea.js');
var CInputGroup = require('./CInputGroup.js');
var CInputGroupText = require('./CInputGroupText.js');

const CFormPlugin = {
    install: (app) => {
        app.component(CForm.CForm.name, CForm.CForm);
        app.component(CFormCheck.CFormCheck.name, CFormCheck.CFormCheck);
        app.component(CFormFeedback.CFormFeedback.name, CFormFeedback.CFormFeedback);
        app.component(CFormFloating.CFormFloating.name, CFormFloating.CFormFloating);
        app.component(CFormInput.CFormInput.name, CFormInput.CFormInput);
        app.component(CFormLabel.CFormLabel.name, CFormLabel.CFormLabel);
        app.component(CFormRange.CFormRange.name, CFormRange.CFormRange);
        app.component(CFormSelect.CFormSelect.name, CFormSelect.CFormSelect);
        app.component(CFormSwitch.CFormSwitch.name, CFormSwitch.CFormSwitch);
        app.component(CFormText.CFormText.name, CFormText.CFormText);
        app.component(CFormTextarea.CFormTextarea.name, CFormTextarea.CFormTextarea);
        app.component(CInputGroup.CInputGroup.name, CInputGroup.CInputGroup);
        app.component(CInputGroupText.CInputGroupText.name, CInputGroupText.CInputGroupText);
    },
};

exports.CForm = CForm.CForm;
exports.CFormCheck = CFormCheck.CFormCheck;
exports.CFormFeedback = CFormFeedback.CFormFeedback;
exports.CFormFloating = CFormFloating.CFormFloating;
exports.CFormInput = CFormInput.CFormInput;
exports.CFormLabel = CFormLabel.CFormLabel;
exports.CFormRange = CFormRange.CFormRange;
exports.CFormSelect = CFormSelect.CFormSelect;
exports.CFormSwitch = CFormSwitch.CFormSwitch;
exports.CFormText = CFormText.CFormText;
exports.CFormTextarea = CFormTextarea.CFormTextarea;
exports.CInputGroup = CInputGroup.CInputGroup;
exports.CInputGroupText = CInputGroupText.CInputGroupText;
exports.CFormPlugin = CFormPlugin;
