// @coreui/vue-pro/dist/cjs/components/nav/index.js
'use strict';

var CNav = require('./CNav.js');
var CNavGroup = require('./CNavGroup.js');
var CNavGroupItems = require('./CNavGroupItems.js');
var CNavItem = require('./CNavItem.js');
var CNavLink = require('./CNavLink.js');
var CNavTitle = require('./CNavTitle.js');

const CNavPlugin = {
    install: (app) => {
        app.component(CNav.CNav.name, CNav.CNav);
        app.component(CNavGroup.CNavGroup.name, CNavGroup.CNavGroup);
        app.component(CNavGroupItems.CNavGroupItems.name, CNavGroupItems.CNavGroupItems);
        app.component(CNavItem.CNavItem.name, CNavItem.CNavItem);
        app.component(CNavLink.CNavLink.name, CNavLink.CNavLink);
        app.component(CNavTitle.CNavTitle.name, CNavTitle.CNavTitle);
    },
};

exports.CNav = CNav.CNav;
exports.CNavGroup = CNavGroup.CNavGroup;
exports.CNavGroupItems = CNavGroupItems.CNavGroupItems;
exports.CNavItem = CNavItem.CNavItem;
exports.CNavLink = CNavLink.CNavLink;
exports.CNavTitle = CNavTitle.CNavTitle;
exports.CNavPlugin = CNavPlugin;
