// @coreui/vue-pro/dist/cjs/components/dropdown/index.js
'use strict';

var CDropdown = require('./CDropdown.js');
var CDropdownItem = require('./CDropdownItem.js');
var CDropdownHeader = require('./CDropdownHeader.js');
var CDropdownDivider = require('./CDropdownDivider.js');
var CDropdownMenu = require('./CDropdownMenu.js');
var CDropdownToggle = require('./CDropdownToggle.js');

const CDropdownPlugin = {
    install: (app) => {
        app.component(CDropdown.CDropdown.name, CDropdown.CDropdown);
        app.component(CDropdownItem.CDropdownItem.name, CDropdownItem.CDropdownItem);
        app.component(CDropdownHeader.CDropdownHeader.name, CDropdownHeader.CDropdownHeader);
        app.component(CDropdownDivider.CDropdownDivider.name, CDropdownDivider.CDropdownDivider);
        app.component(CDropdownMenu.CDropdownMenu.name, CDropdownMenu.CDropdownMenu);
        app.component(CDropdownToggle.CDropdownToggle.name, CDropdownToggle.CDropdownToggle);
    },
};

exports.CDropdown = CDropdown.CDropdown;
exports.CDropdownItem = CDropdownItem.CDropdownItem;
exports.CDropdownHeader = CDropdownHeader.CDropdownHeader;
exports.CDropdownDivider = CDropdownDivider.CDropdownDivider;
exports.CDropdownMenu = CDropdownMenu.CDropdownMenu;
exports.CDropdownToggle = CDropdownToggle.CDropdownToggle;
exports.CDropdownPlugin = CDropdownPlugin;
