// @coreui/vue-pro/dist/cjs/components/widgets/index.js
'use strict';

var CWidgetStatsA = require('./CWidgetStatsA.js');
var CWidgetStatsB = require('./CWidgetStatsB.js');
var CWidgetStatsC = require('./CWidgetStatsC.js');
var CWidgetStatsD = require('./CWidgetStatsD.js');
var CWidgetStatsE = require('./CWidgetStatsE.js');
var CWidgetStatsF = require('./CWidgetStatsF.js');

const CWidgetsStatsPlugin = {
    install: (app) => {
        app.component(CWidgetStatsA.CWidgetStatsA.name, CWidgetStatsA.CWidgetStatsA);
        app.component(CWidgetStatsB.CWidgetStatsB.name, CWidgetStatsB.CWidgetStatsB);
        app.component(CWidgetStatsC.CWidgetStatsC.name, CWidgetStatsC.CWidgetStatsC);
        app.component(CWidgetStatsD.CWidgetStatsD.name, CWidgetStatsD.CWidgetStatsD);
        app.component(CWidgetStatsE.CWidgetStatsE.name, CWidgetStatsE.CWidgetStatsE);
        app.component(CWidgetStatsF.CWidgetStatsF.name, CWidgetStatsF.CWidgetStatsF);
    },
};

exports.CWidgetStatsA = CWidgetStatsA.CWidgetStatsA;
exports.CWidgetStatsB = CWidgetStatsB.CWidgetStatsB;
exports.CWidgetStatsC = CWidgetStatsC.CWidgetStatsC;
exports.CWidgetStatsD = CWidgetStatsD.CWidgetStatsD;
exports.CWidgetStatsE = CWidgetStatsE.CWidgetStatsE;
exports.CWidgetStatsF = CWidgetStatsF.CWidgetStatsF;
exports.CWidgetsStatsPlugin = CWidgetsStatsPlugin;
