// @coreui/vue-pro/dist/cjs/components/table/CTableRow.js
'use strict';

var vue = require('vue');
var props = require('../../props.js');

const CTableRow = vue.defineComponent({
    name: 'CTableRow',
    props: {
        /**
         * Highlight a table row or cell..
         */
        active: Boolean,
        /**
         * Set the vertical aligment.
         *
         * @values 'bottom', 'middle', 'top'
         */
        align: {
            type: String,
            validator: (value) => {
                return ['bottom', 'middle', 'top'].includes(value);
            },
        },
        /**
         * Sets the color context of the component to one of CoreUI’s themed colors.
         *
         * @values  'primary', 'secondary', 'success', 'danger', 'warning', 'info', 'dark', 'light', string
         */
        color: props.Color,
    },
    setup(props, { slots }) {
        return () => vue.h('tr', {
            class: [
                {
                    [`align-${props.align}`]: props.align,
                    'table-active': props.active,
                    [`table-${props.color}`]: props.color,
                },
            ],
        }, slots.default && slots.default());
    },
});

exports.CTableRow = CTableRow;
