// @coreui/vue-pro/dist/cjs/components/table/index.js
'use strict';

var CTable = require('./CTable.js');
var CTableBody = require('./CTableBody.js');
var CTableCaption = require('./CTableCaption.js');
var CTableDataCell = require('./CTableDataCell.js');
var CTableFoot = require('./CTableFoot.js');
var CTableHead = require('./CTableHead.js');
var CTableHeaderCell = require('./CTableHeaderCell.js');
var CTableRow = require('./CTableRow.js');

const CTablePlugin = {
    install: (app) => {
        app.component(CTable.CTable.name, CTable.CTable);
        app.component(CTableBody.CTableBody.name, CTableBody.CTableBody);
        app.component(CTableCaption.CTableCaption.name, CTableCaption.CTableCaption);
        app.component(CTableDataCell.CTableDataCell.name, CTableDataCell.CTableDataCell);
        app.component(CTableFoot.CTableFoot.name, CTableFoot.CTableFoot);
        app.component(CTableHead.CTableHead.name, CTableHead.CTableHead);
        app.component(CTableHeaderCell.CTableHeaderCell.name, CTableHeaderCell.CTableHeaderCell);
        app.component(CTableRow.CTableRow.name, CTableRow.CTableRow);
    },
};

exports.CTable = CTable.CTable;
exports.CTableBody = CTableBody.CTableBody;
exports.CTableCaption = CTableCaption.CTableCaption;
exports.CTableDataCell = CTableDataCell.CTableDataCell;
exports.CTableFoot = CTableFoot.CTableFoot;
exports.CTableHead = CTableHead.CTableHead;
exports.CTableHeaderCell = CTableHeaderCell.CTableHeaderCell;
exports.CTableRow = CTableRow.CTableRow;
exports.CTablePlugin = CTablePlugin;
