// @coreui/vue-pro/dist/cjs/components/header/index.js
'use strict';

var CHeader = require('./CHeader.js');
var CHeaderBrand = require('./CHeaderBrand.js');
var CHeaderDivider = require('./CHeaderDivider.js');
var CHeaderNav = require('./CHeaderNav.js');
var CHeaderText = require('./CHeaderText.js');
var CHeaderToggler = require('./CHeaderToggler.js');

const CHeaderPlugin = {
    install: (app) => {
        app.component(CHeader.CHeader.name, CHeader.CHeader);
        app.component(CHeaderBrand.CHeaderBrand.name, CHeaderBrand.CHeaderBrand);
        app.component(CHeaderDivider.CHeaderDivider.name, CHeaderDivider.CHeaderDivider);
        app.component(CHeaderNav.CHeaderNav.name, CHeaderNav.CHeaderNav);
        app.component(CHeaderText.CHeaderText.name, CHeaderText.CHeaderText);
        app.component(CHeaderToggler.CHeaderToggler.name, CHeaderToggler.CHeaderToggler);
    },
};

exports.CHeader = CHeader.CHeader;
exports.CHeaderBrand = CHeaderBrand.CHeaderBrand;
exports.CHeaderDivider = CHeaderDivider.CHeaderDivider;
exports.CHeaderNav = CHeaderNav.CHeaderNav;
exports.CHeaderText = CHeaderText.CHeaderText;
exports.CHeaderToggler = CHeaderToggler.CHeaderToggler;
exports.CHeaderPlugin = CHeaderPlugin;
