// @coreui/vue-pro/dist/cjs/components/sidebar/index.js
'use strict';

var CSidebar = require('./CSidebar.js');
var CSidebarBrand = require('./CSidebarBrand.js');
var CSidebarFooter = require('./CSidebarFooter.js');
var CSidebarHeader = require('./CSidebarHeader.js');
var CSidebarNav = require('./CSidebarNav.js');
var CSidebarToggler = require('./CSidebarToggler.js');

const CSidebarPlugin = {
    install: (app) => {
        app.component(CSidebar.CSidebar.name, CSidebar.CSidebar);
        app.component(CSidebarBrand.CSidebarBrand.name, CSidebarBrand.CSidebarBrand);
        app.component(CSidebarFooter.CSidebarFooter.name, CSidebarFooter.CSidebarFooter);
        app.component(CSidebarHeader.CSidebarHeader.name, CSidebarHeader.CSidebarHeader);
        app.component(CSidebarNav.CSidebarNav.name, CSidebarNav.CSidebarNav);
        app.component(CSidebarToggler.CSidebarToggler.name, CSidebarToggler.CSidebarToggler);
    },
};

exports.CSidebar = CSidebar.CSidebar;
exports.CSidebarBrand = CSidebarBrand.CSidebarBrand;
exports.CSidebarFooter = CSidebarFooter.CSidebarFooter;
exports.CSidebarHeader = CSidebarHeader.CSidebarHeader;
exports.CSidebarNav = CSidebarNav.CSidebarNav;
exports.CSidebarToggler = CSidebarToggler.CSidebarToggler;
exports.CSidebarPlugin = CSidebarPlugin;
