// @coreui/vue-pro/dist/cjs/components/card/index.js
'use strict';

var CCard = require('./CCard.js');
var CCardBody = require('./CCardBody.js');
var CCardFooter = require('./CCardFooter.js');
var CCardGroup = require('./CCardGroup.js');
var CCardHeader = require('./CCardHeader.js');
var CCardImage = require('./CCardImage.js');
var CCardImageOverlay = require('./CCardImageOverlay.js');
var CCardLink = require('./CCardLink.js');
var CCardSubtitle = require('./CCardSubtitle.js');
var CCardText = require('./CCardText.js');
var CCardTitle = require('./CCardTitle.js');

const CCardPlugin = {
    install: (app) => {
        app.component(CCard.CCard.name, CCard.CCard);
        app.component(CCardBody.CCardBody.name, CCardBody.CCardBody);
        app.component(CCardFooter.CCardFooter.name, CCardFooter.CCardFooter);
        app.component(CCardGroup.CCardGroup.name, CCardGroup.CCardGroup);
        app.component(CCardHeader.CCardHeader.name, CCardHeader.CCardHeader);
        app.component(CCardImage.CCardImage.name, CCardImage.CCardImage);
        app.component(CCardImageOverlay.CCardImageOverlay.name, CCardImageOverlay.CCardImageOverlay);
        app.component(CCardLink.CCardLink.name, CCardLink.CCardLink);
        app.component(CCardSubtitle.CCardSubtitle.name, CCardSubtitle.CCardSubtitle);
        app.component(CCardText.CCardText.name, CCardText.CCardText);
        app.component(CCardTitle.CCardTitle.name, CCardTitle.CCardTitle);
    },
};

exports.CCard = CCard.CCard;
exports.CCardBody = CCardBody.CCardBody;
exports.CCardFooter = CCardFooter.CCardFooter;
exports.CCardGroup = CCardGroup.CCardGroup;
exports.CCardHeader = CCardHeader.CCardHeader;
exports.CCardImage = CCardImage.CCardImage;
exports.CCardImageOverlay = CCardImageOverlay.CCardImageOverlay;
exports.CCardLink = CCardLink.CCardLink;
exports.CCardSubtitle = CCardSubtitle.CCardSubtitle;
exports.CCardText = CCardText.CCardText;
exports.CCardTitle = CCardTitle.CCardTitle;
exports.CCardPlugin = CCardPlugin;
